src/posix/platform/misc.cpp

```diff
@@ -107,6 +107,31 @@ int SocketWithCloseExec(int aDomain, int aType, int aProtocol)
     return fd;
 }
 
+int NonBlockSocketWithCloseExec(int aDomain, int aType, int aProtocol)
+{
+    int rval = 0;
+    int fd   = -1;
+
+#ifdef __APPLE__
+    VerifyOrExit((fd = socket(aDomain, aType, aProtocol)) != -1, perror("socket(SOCK_CLOEXEC)"));
+
+    VerifyOrExit((rval = fcntl(fd, F_GETFD, 0)) != -1, perror("fcntl(F_GETFD)"));
+    VerifyOrExit((rval = fcntl(fd, F_SETFD, rval | FD_CLOEXEC | O_NONBLOCK)) != -1, perror("fcntl(F_SETFD)"));
+#else
+    VerifyOrExit((fd = socket(aDomain, aType | SOCK_CLOEXEC | SOCK_NONBLOCK, aProtocol)) != -1,
+                 perror("socket(SOCK_CLOEXEC)"));
+#endif
+
+exit:
+    if (rval == -1 && fd != -1)
```

Comment:
nit: `fd` must not be -1 if `rval == -1` ?
```suggestion
    if (rval == -1)
```